watertap/costing/costing_base.py

```diff
@@ -0,0 +1,239 @@
+
+    def add_specific_electrical_carbon_intensity(
+        self, flow_rate, name="specific_electrical_carbon_intensity"
+    ):
+        """
+        Add specific electrical carbon intensity (kg_CO2eq/m**3) to costing block.
+        Args:
+            flow_rate - flow rate of water (volumetric) to be used in
+                        calculating specific electrical carbon intensity
+            name (optional) - the name of the Expression for the specific
+                              carbon intensity (default: specific_electrical_carbon_intensity)
+        """
+
+        self.add_component(
+            name,
+            pyo.Expression(
+                expr=self.aggregate_flow_electricity
+                * self.electrical_carbon_intensity
+                / pyo.units.convert(
+                    flow_rate, to_units=pyo.units.m**3 / pyo.units.hr
+                ),
+                doc=f"Specific electrical carbon intensity based on flow {flow_rate.name}",
+            ),
+        )
+
+    def _build_common_global_params(self):
+        """
+        Build the global parameters common to WaterTAP Costing Packages.
+        The currency units should already be registered.
+
+        The derived class should define the capital_recovery_factor.
+        """
+
+        self.utilization_factor = pyo.Var(
+            initialize=0.9,
+            doc="Plant capacity utilization [fraction of uptime]",
+            units=pyo.units.dimensionless,
+        )
+
+        self.electricity_cost = pyo.Var(
+            initialize=0.07,
+            doc="Electricity cost",
+            units=pyo.units.USD_2018 / pyo.units.kWh,
+        )
+        self.defined_flows["electricity"] = self.electricity_cost
+
+        self.electrical_carbon_intensity = pyo.Var(
+            initialize=0.475,
+            doc="Grid carbon intensity [kgCO2_eq/kWh]",
+            units=pyo.units.kg / pyo.units.kWh,
+        )
+
+        self.capital_recovery_factor = pyo.Expression(
+            expr=0,
+            doc="Capital annualization factor [fraction of investment cost/year]",
+        )
+
+        self.TPEC = pyo.Var(
+            initialize=3.4,
+            doc="Total Purchased Equipment Cost (TPEC)",
+            units=pyo.units.dimensionless,
+        )
+
+        self.TIC = pyo.Var(
+            initialize=1.65,
+            doc="Total Installed Cost (TIC)",
+            units=pyo.units.dimensionless,
+        )
+
+        self.fix_all_vars()
+
+    def _get_costing_method_for(self, unit_model):
+        """
+        Allow the unit model to register its default costing method,
+        either through an attribute named "default_costing_method"
+        or by naming the default costing method "default_costing_method"
+        """
+        if hasattr(unit_model, "default_costing_method"):
+            return unit_model.default_costing_method
+        return super()._get_costing_method_for(unit_model)
+
+    def register_flow_type(self, flow_type, cost):
```

Comment:
Can you explain what this is doing and why it is necessary? No issue with the code itself, just what it is doing in the larger framework here and why are units involved.

Reply:
Ah, this goes back to something I think you found a while ago.

The IDAES costing framework will create a new `Var` in `register_flow_type`, unless there already is a component on the costing package named `flow_type + "_cost"`.

But this creates issues if you want to use mutable parameters or other variables to define the `cost`, then changes to those won't be reflected. There's a bit more discussion on the issue we're side-stepping here: https://github.com/IDAES/idaes-pse/pull/1014#issuecomment-1309527271.

Reply:
As for the units, it has to do with the fact that Pyomo treats units as expressions. So, e.g., we want `register_flow_type( "foo", 42.0 * pyo.units.USD_2020 / pyo.units.m )` to go ahead and create a new variable called `foo_cost` whose value is `42.0 * pyo.units.USD_2020 / pyo.units.m`.

But if I define a parameter or variable named `bar_base_cost` and another parameter called `bar_purity`, then `register_flow_type( "bar", bar_base_cost / bar_purity)`, then we will *not* create a new `Var`, but instead create a new `Expression` called `bar_cost` with `bar_cost.expr = bar_base_cost / bar_purity`. It is assumed that if the user wants to do parameter sweeps over the bar_cost the user will modify `bar_base_cost`.


There is a proposed fix for some of this ugliness, but it hasn't been a priority for IDAES/Pyomo: https://github.com/IDAES/idaes-pse/pull/1014#issuecomment-1332709400